datasafe-simple-adapter/datasafe-simple-adapter-impl/src/main/java/de/adorsys/datasafe/simple/adapter/impl/SimpleDatasafeServiceImpl.java

```diff
@@ -137,31 +132,15 @@ public SimpleDatasafeServiceImpl(DFSCredentials dfsCredentials, MutableEncryptio
             );
             this.systemRoot = URI.create(S3_PREFIX + amazonS3DFSCredentials.getRootBucket());
 
-            customlyBuiltDatasafeServices = DaggerLegacyDatasafeService.builder()
+            customlyBuiltDatasafeServices = DaggerSwitchableDatasafeServices.builder()
```

Comment:
Instead of duplicating twice i suggest to extract builder:
```
public SimpleDatasafeServiceImpl(DFSCredentials dfsCredentials, MutableEncryptionConfig config) {

        SwitchableDatasafeServices.Builder serviceBuilder = DaggerSwitchableDatasafeServices.builder()
                .config(new DefaultDFSConfig(systemRoot, universalReadStorePassword))
                .encryption(config.toEncryptionConfig());

        if (dfsCredentials instanceof FilesystemDFSCredentials) {
            FilesystemDFSCredentials filesystemDFSCredentials = (FilesystemDFSCredentials) dfsCredentials;
            LogStringFrame lsf = new LogStringFrame();
            lsf.add("FILESYSTEM");
            lsf.add("root bucket     : " + filesystemDFSCredentials.getRoot());
            lsf.add("path encryption : " + SwitchablePathEncryptionImpl.checkIsPathEncryptionToUse());
            log.info(lsf.toString());
            this.systemRoot = FileSystems.getDefault().getPath(filesystemDFSCredentials.getRoot()).toAbsolutePath().toUri();
            storageService = new FileSystemStorageService(FileSystems.getDefault().getPath(filesystemDFSCredentials.getRoot()));
            log.info("build DFS to FILESYSTEM with root " + filesystemDFSCredentials.getRoot());
        }
        if (dfsCredentials instanceof AmazonS3DFSCredentials) {
            AmazonS3DFSCredentials amazonS3DFSCredentials = (AmazonS3DFSCredentials) dfsCredentials;
            LogStringFrame lsf = new LogStringFrame();
            lsf.add("AMAZON S3");
            lsf.add("root bucket     : " + amazonS3DFSCredentials.getRootBucket());
            lsf.add("url             : " + amazonS3DFSCredentials.getUrl());
            lsf.add("region          : " + amazonS3DFSCredentials.getRegion());
            lsf.add("path encryption : " + SwitchablePathEncryptionImpl.checkIsPathEncryptionToUse());
            lsf.add("no https        : " + amazonS3DFSCredentials.isNoHttps());
            lsf.add("threadpool size : " + amazonS3DFSCredentials.getThreadPoolSize());
            log.info(lsf.toString());
            AmazonS3ClientBuilder amazonS3ClientBuilder = AmazonS3ClientBuilder.standard()
                    .withCredentials(
                            new AWSStaticCredentialsProvider(
                                    new BasicAWSCredentials(
                                            amazonS3DFSCredentials.getAccessKey(),
                                            amazonS3DFSCredentials.getSecretKey()))
                    )
                    .enablePathStyleAccess();

            boolean useEndpoint = !amazonS3DFSCredentials.getUrl().equals(AMAZON_URL)
                    && !amazonS3DFSCredentials.getUrl().startsWith(S3_PREFIX);
            if (useEndpoint) {
                AwsClientBuilder.EndpointConfiguration endpoint = new AwsClientBuilder.EndpointConfiguration(
                        amazonS3DFSCredentials.getUrl(),
                        amazonS3DFSCredentials.getRegion()
                );
                amazonS3ClientBuilder.withEndpointConfiguration(endpoint);
            } else {
                amazonS3ClientBuilder.withRegion(amazonS3DFSCredentials.getRegion());
            }

            if (amazonS3DFSCredentials.isNoHttps()) {
                log.info("Creating S3 client without https");
                ClientConfiguration clientConfig = new ClientConfiguration();
                clientConfig.setProtocol(Protocol.HTTP);
                clientConfig.disableSocketProxy();
                amazonS3ClientBuilder.withClientConfiguration(clientConfig);
            }

            AmazonS3 amazons3 = amazonS3ClientBuilder.build();

            if (!amazons3.doesBucketExistV2(amazonS3DFSCredentials.getContainer())) {
                amazons3.createBucket(amazonS3DFSCredentials.getContainer());
            }
            storageService = new S3StorageService(
                    amazons3,
                    amazonS3DFSCredentials.getContainer(),
                    ExecutorServiceUtil
                            .submitterExecutesOnStarvationExecutingService(
                                    amazonS3DFSCredentials.getThreadPoolSize(),
                                    amazonS3DFSCredentials.getQueueSize()
                            )
            );
            this.systemRoot = URI.create(S3_PREFIX + amazonS3DFSCredentials.getRootBucket());
            log.info("build DFS to S3 with root " + amazonS3DFSCredentials.getRootBucket() + " and url " + amazonS3DFSCredentials.getUrl());
        }

        customlyBuiltDatasafeServices = serviceBuilder
                .storage(getStorageService())
                .build();
    }
```

Reply:
Also handling different storage types could be done in separate functions

```
 if (dfsCredentials instanceof FilesystemDFSCredentials) {
useFileSystem();
}
....
void useFileSystem() {
filesystemDFSCredentials filesystemDFSCredentials = (FilesystemDFSCredentials) dfsCredentials;
            LogStringFrame lsf = new LogStringFrame();
            lsf.add("FILESYSTEM");
            lsf.add("root bucket     : " + filesystemDFSCredentials.getRoot());
            lsf.add("path encryption : " + SwitchablePathEncryptionImpl.checkIsPathEncryptionToUse());
            log.info(lsf.toString());
            this.systemRoot = FileSystems.getDefault().getPath(filesystemDFSCredentials.getRoot()).toAbsolutePath().toUri();
            storageService = new FileSystemStorageService(FileSystems.getDefault().getPath(filesystemDFSCredentials.getRoot()));
            log.info("build DFS to FILESYSTEM with root " + filesystemDFSCredentials.getRoot());
}
```

Reply:
Did so. But as a strongly dislike methods which get input an change some members unvisibly for caller I made methods static and return results to caller.